### ISSUE TITLE
NVDA 2020.3 beta 4 crashes constantly with actilino braille display

### ISSUE THREAD

**Issue by manish10** [nvaccess/nvda#11723]
### Steps to reproduce:
Start NVDA with Actilino braille display.

### Actual behavior:
NVDA constantly crashes every few minutes
### Expected behavior:
does not crash
#### Have you tried any other versions of NVDA? If so, please report their behaviors. 
This was not happening in beta 3.

#### If addons are disabled, is your problem still occuring?
did not try.
#### Did you try to run the COM registry fixing tool in NVDA menu / tools?
no
[nvda.log](https://github.com/nvaccess/nvda/files/5322727/nvda.log)

**Comment by MarcoZehe** [comment]
CC @FelixGruetzmacher for awareness. FWIW, the log doesn't contain any useful info other than the system suddenly wants to unregister the hidden driver window. I currently only have an older BrailleStar 40 here and cannot reproduce any problems.

**Comment by FelixGruetzmacher** [comment]
How do these crashes manifest? Does NVDA simply stop speaking and Brailling? Do you then restart it with ctrl+alt+n? Does it restart itself? What are you typically doing when it happens? Was this restricted to Outlook?

**Comment by manish10** [comment]
It was happening consistently with outlook in the compose new mail as I was
typing.
NVDA would stop working and I would need to bring it back up with "alt ctrl
N".
I have it running right now also and it doesn't seem to crash any more. So,
seems to be some intermittent behavior
.

Regards,
Manish
http://iaccessible.net <http://iaccessible.blogspot.com>

**Comment by Afik-souffir998** [comment]
Its crashing also with QBrailleXl from HIMS for unknown reason.


Sent from my Iphone

On 5 Oct 2020, at 22:31, manish10 <notifications@github.com> wrote:

﻿

It was happening consistently with outlook in the compose new mail as I was
typing.
NVDA would stop working and I would need to bring it back up with "alt ctrl
N".
I have it running right now also and it doesn't seem to crash any more. So,
seems to be some intermittent behavior
.

Regards,
Manish
http://iaccessible.net <http://iaccessible.blogspot.com>

**Comment by MarcoZehe** [comment]
Are the symptoms possibly similar to what I described in #11576 ? E. g., if you lose speech and braille the next time, can you wait a few seconds and listen to the error sound (two quite dissonant descending piano cords)? If you hear them, and then about 20 seconds later again, you might be running into the same issue. Restart NVDA as usual, then look in your %TEMP% folder for a nvda-old.log file and attach it to this issue, please. That is the log file from the session that actually froze/crashed once you have NVDA restarted.

**Comment by leonardder** [comment]
How is this with RC1?

**Comment by manish10** [comment]
It has not crashed so far since the RC1. I have been running it all day
today with actilino.

Regards,
Manish
http://iaccessible.net <http://iaccessible.blogspot.com>

**Comment by FelixGruetzmacher** [comment]
What made you originally assume the crashes were Braille-related?

**Comment by manish10** [comment]
Because the crashes stopped happening if I unplugged the braille display
and started happening again when I plugged it back in.

Regards,
Manish
http://iaccessible.net <http://iaccessible.blogspot.com>

**Comment by feerrenrut** [comment]
Relevant part of the log:

```
INFO - brailleDisplayDrivers.handyTech.BrailleDisplayDriver.__init__ (10:13:51.358) - braille._BgThread (15152):
Found Actilino connected via serial (COM4)
INFO - braille.BrailleHandler.setDisplayByName (10:13:51.387) - braille._BgThread (15152):
Loaded braille display driver handyTech, current display has 16 cells.
ERROR - windowUtils.CustomWindow.destroy (10:13:51.835) - MainThread (2196):
Error destroying window for InvisibleDriverWindow
OSError: [WinError 1400] Invalid window handle.
ERROR - windowUtils.CustomWindow.destroy (10:13:51.904) - MainThread (2196):
Error unregistering window class for InvisibleDriverWindow
OSError: [WinError 87] The parameter is incorrect.
INFO - braille.BrailleHandler.setDisplayByName (10:13:52.281) - braille._BgThread (15152):
Loaded braille display driver noBraille, current display has 0 cells.
ERROR - windowUtils.CustomWindow.destroy (10:13:52.281) - braille._BgThread (15152):
Error destroying window for InvisibleDriverWindow
OSError: [WinError 1400] Invalid window handle.
ERROR - windowUtils.CustomWindow.destroy (10:13:52.320) - braille._BgThread (15152):
Error unregistering window class for InvisibleDriverWindow
OSError: [WinError 87] The parameter is incorrect.
INFO - braille.BrailleHandler.setDisplayByName (10:13:52.355) - braille._BgThread (15152):
Loaded braille display driver noBraille, current display has 0 cells.
```

**Comment by feerrenrut** [comment]
If this is now working, which it seems to be given the comment from the 9th of October, I think we can close the issue.

**Comment by leonardder** [comment]
This error sounds very familiar. In fact, I can reproduce it pretty easily on my system.

1. Attach a Modular evolution display
2. Enable braille display auto detection
3. When the display is connected, set braille display to no braille

Likely has to do with window initialization not being thread safe.

**Comment by feerrenrut** [comment]
Ah, in that case, this sounds like something that should be fixed. Thanks for the STR

**Comment by leonardder** [comment]
I quickly investigated this and I"m afraid that the custom message 
window is full of buggyness. A major reason for this is that it is bound 
to the instance of the driver, and when switching displays, two 
instances of the driver can exist at the same time, though it should be 
a very short time. This could happen to switch from automatic to handy 
tech, from automatic to automatic, etc. Also, the window is destroyed 
multiple times when terminating the driver, once in terminate and once 
in the weak reference to the driver on the window itself.

**Comment by leonardder** [comment]
I'm pretty sure I fixed this at some point, namely in #12618. Closing this as complete.
